openvdb/Types.h

```diff
@@ -337,6 +337,167 @@ struct CanConvertType<ValueMask, T> { enum {value = CanConvertType<bool, T>::val
 
 
+/// @brief CopyConstness<T1, T2>::Type is either <tt>const T2</tt>
+/// or @c T2 with no @c const qualifier, depending on whether @c T1 is @c const.
+/// @details For example,
+/// - CopyConstness<int, int>::Type is @c int
+/// - CopyConstness<int, const int>::Type is @c int
+/// - CopyConstness<const int, int>::Type is <tt>const int</tt>
+/// - CopyConstness<const int, const int>::Type is <tt>const int</tt>
+template<typename FromType, typename ToType> struct CopyConstness {
+    using Type = typename std::remove_const<ToType>::type;
+};
+
+/// @cond OPENVDB_TYPES_INTERNAL
+template<typename FromType, typename ToType> struct CopyConstness<const FromType, ToType> {
+    using Type = const ToType;
+};
+/// @endcond
+
+
+////////////////////////////////////////
+
+
+/// @cond OPENVDB_TYPES_INTERNAL
+
+template<typename... Ts> struct TypeList; // forward declaration
+
+namespace internal {
+
+// Implementation details of TypeList
+
+template<typename ListT, typename... Ts> struct TSAppendImpl;
+
+// Append zero or more types.
+template<typename... Ts, typename... OtherTs>
+struct TSAppendImpl<TypeList<Ts...>, OtherTs...> {
+    using type = TypeList<Ts..., OtherTs...>;
+};
+
+// Append another TypeList's members.
+template<typename... Ts, typename... OtherTs>
+struct TSAppendImpl<TypeList<Ts...>, TypeList<OtherTs...>> {
+    using type = TypeList<Ts..., OtherTs...>;
+};
+
+
+// Remove all occurrences of type T.
+template<typename ListT, typename T> struct TSEraseImpl;
+
+// TypeList<>::Erase<int> = TypeList<>
+template<typename T>
+struct TSEraseImpl<TypeList<>, T> { using type = TypeList<>; };
+
+// TypeList<int, char, ...>::Erase<int> = TypeList<char, ...>::Erase<int>
+template<typename... Ts, typename T>
+struct TSEraseImpl<TypeList<T, Ts...>, T> {
+    using type = typename TSEraseImpl<TypeList<Ts...>, T>::type;
+};
+
+// TypeList<float, int, char...>::Erase<int> =
+//     TypeList<float>::Append<TypeList<int, char...>::Erase<int>>
+template<typename T2, typename... Ts, typename T>
+struct TSEraseImpl<TypeList<T2, Ts...>, T> {
+    using type = typename TSAppendImpl<TypeList<T2>,
+        typename TSEraseImpl<TypeList<Ts...>, T>::type>::type;
+};
+
+
+template<typename ListT, typename... Ts> struct TSRemoveImpl;
+
+template<typename ListT>
+struct TSRemoveImpl<ListT> { using type = ListT; };
+
+// Remove one or more types.
+template<typename ListT, typename T, typename... Ts>
+struct TSRemoveImpl<ListT, T, Ts...> {
+    using type = typename TSRemoveImpl<typename TSEraseImpl<ListT, T>::type, Ts...>::type;
+};
+
+// Remove the members of another TypeList.
+template<typename ListT, typename... Ts>
+struct TSRemoveImpl<ListT, TypeList<Ts...>> {
+    using type = typename TSRemoveImpl<ListT, Ts...>::type;
+};
+
+
+template<typename OpT> inline void TSForEachImpl(OpT) {}
+template<typename OpT, typename T, typename... Ts>
+inline void TSForEachImpl(OpT op) { op(T()); TSForEachImpl<OpT, Ts...>(op); }
+
+} // namespace internal
```

Comment:
These could go all into TypeList instead of an internal namespace (and could then become private). Depends how you feel about nested structs

Reply:
ICC 15 (at least) seems to have an issue with parameter pack expansion in nested, templated classes.  I don't see a good way to work around the problem, so I'll leave this as is for now.